File: java/src/hashing/InternationalPhoneBookDynamicArray.java
package hashing;

public class InternationalPhoneBookDynamicArray {
    private Pair[] arr = new Pair[1];

    private int size = 0;

    private int maxSize = 1;

    private class Pair {
        private String number;

        private String name;

        public Pair(String number, String name) {
            this.number = number;
            this.name = name;
        }
    }

    private int binarySearch(long number, int st, int ed) {
        if (st == ed) {
            if (this.getLong(this.arr[st].number) < number)
                return st + 1;
            else
                return st;
        }

        int mid = st + Math.floorDiv(ed - st, 2);
        long midVal = this.getLong(this.arr[mid].number);

        if (midVal == number)
            return mid;
        else if (midVal < number)
            return this.binarySearch(number, mid + 1, ed);
        else
            return this.binarySearch(number, st, mid - 1);
    }

    private long getLong(String number) {
        return Long.parseLong(number.replaceAll("\\D", ""));
    }

    public void setName(String number, String name) {
        if (this.size == this.maxSize) {
            int newMaxSize = this.maxSize * 2;
            Pair[] newArr = new Pair[newMaxSize];

            for (int i = 0 ; i < this.size ; i++)
                newArr[i] = this.arr[i];

            this.arr = newArr;
            this.maxSize = newMaxSize;
        }

        Pair pair = new Pair(number, name);
        if (this.size == 0) {
            this.arr[this.size] = pair;
        } else {
            int idx = this.binarySearch(this.getLong(number), 0, this.size - 1);
            for (int i = this.size - 1 ; i > idx - 1 ; i--)
                this.arr[i + 1] = this.arr[i];
            this.arr[idx] = pair;
        }
        this.size += 1;
    }

    public String getName(String number) {
        Pair pair = this.arr[this.binarySearch(this.getLong(number), 0, this.size - 1)];

        if (pair != null && pair.number == number)
            return pair.name;
        else
            return null;
    }
}

File: python/hashing/internationalPhoneBookDynamicArray.py
import re

class InternationalPhoneBookDynamicArray:
    def __init__(self):
        self.arr = [None] * 1
        self.size = 0
        self.maxSize = 1

    class Pair:
        def __init__(self, number, name):
            self.number = number
            self.name = name

    def _binarySearch(self, number, st, ed):
        if st == ed:
            if self._getInt(self.arr[st].number) < number:
                return st + 1
            else:
                return st

        mid = st + (ed - st) // 2
        val = self._getInt(self.arr[mid].number)
        
        if val == number:
            return mid
        elif val < number:
            return self._binarySearch(number, mid + 1, ed)
        else:
            return self._binarySearch(number, st, mid - 1)

    def _getInt(self, number):
        return re.sub('\D', '', number)

    def setName(self, number, name):
        # resize
        if self.size == self.maxSize:
            newMaxSize = self.maxSize * 2
            newArr = [None] * newMaxSize
            
            for i in range(self.maxSize):
                newArr[i] = self.arr[i]
            
            self.arr = newArr
            self.maxSize = newMaxSize
        
        # append
        pair = self.Pair(number, name)
        if self.size == 0:
            self.arr[self.size] = pair
        else:
            idx = self._binarySearch(self._getInt(number), 0, self.size - 1)
            for i in range(self.size - 1, idx - 1):
                self.arr[i + 1] = self.arr[i]
            self.arr[idx] = pair

        self.size += 1

    def getName(self, number):
        idx = self._binarySearch(self._getInt(number), 0, self.size - 1)
        val = self.arr[idx]

        if val is not None and self.arr[idx].number == number:
            return self.arr[idx].name
        else:
            return None
